// @mysten/walrus/dist/utils/object-loader.mjs
import { Field } from "./bcs.mjs";
import { deriveDynamicFieldID } from "@mysten/sui/utils";
import { pureBcsSchemaFromTypeName } from "@mysten/sui/bcs";
import DataLoader from "dataloader";

//#region src/utils/object-loader.ts
var SuiObjectDataLoader = class extends DataLoader {
	#dynamicFieldCache = /* @__PURE__ */ new Map();
	constructor(suiClient) {
		super(async (ids) => {
			const { objects } = await suiClient.core.getObjects({
				objectIds: ids,
				include: { content: true }
			});
			return objects;
		});
	}
	async load(id, schema) {
		const data = await super.load(id);
		if (schema) return schema.parse(data.content);
		return data;
	}
	async loadMany(ids, schema) {
		const data = await super.loadMany(ids);
		if (!schema) return data;
		return data.map((d) => {
			if (d instanceof Error) return d;
			return schema.parse(d.content);
		});
	}
	async loadManyOrThrow(ids, schema) {
		const data = await this.loadMany(ids, schema);
		for (const d of data) if (d instanceof Error) throw d;
		return data;
	}
	clearAll() {
		this.#dynamicFieldCache.clear();
		return super.clearAll();
	}
	clear(key) {
		this.#dynamicFieldCache.delete(key);
		return super.clear(key);
	}
	async loadFieldObject(parent, name, type) {
		const schema = pureBcsSchemaFromTypeName(name.type);
		const id = deriveDynamicFieldID(parent, "u64", schema.serialize(name.value).toBytes());
		return (await this.load(id, Field(schema, type))).value;
	}
};

//#endregion
export { SuiObjectDataLoader };
