// @mysten/walrus/dist/contracts/walrus/auth.mjs
import { MoveEnum } from "../utils/index.mjs";
import { bcs } from "@mysten/sui/bcs";

//#region src/contracts/walrus/auth.ts
/**************************************************************
* THIS FILE IS GENERATED AND SHOULD NOT BE MANUALLY MODIFIED *
**************************************************************/
const $moduleName = "@local-pkg/walrus::auth";
/**
* Authentication for either a sender or an object. Unlike the `Authorized` type,
* it cannot be stored and must be used or ignored in the same transaction.
*/
const Authenticated = new MoveEnum({
	name: `${$moduleName}::Authenticated`,
	fields: {
		Sender: bcs.Address,
		Object: bcs.Address
	}
});
/**
* Defines the ways to authorize an action. It can be either an address - checked
* with `ctx.sender()`, - or an object - checked with `object::id(..)`.
*/
const Authorized = new MoveEnum({
	name: `${$moduleName}::Authorized`,
	fields: {
		Address: bcs.Address,
		ObjectID: bcs.Address
	}
});

//#endregion
export { Authorized };
