// @mysten/walrus/dist/contracts/walrus/blob.mjs
import { MoveStruct, normalizeMoveArguments } from "../utils/index.mjs";
import { Storage } from "./storage_resource.mjs";
import { bcs } from "@mysten/sui/bcs";

//#region src/contracts/walrus/blob.ts
/**************************************************************
* THIS FILE IS GENERATED AND SHOULD NOT BE MANUALLY MODIFIED *
**************************************************************/
const $moduleName = "@local-pkg/walrus::blob";
const Blob = new MoveStruct({
	name: `${$moduleName}::Blob`,
	fields: {
		id: bcs.Address,
		registered_epoch: bcs.u32(),
		blob_id: bcs.u256(),
		size: bcs.u64(),
		encoding_type: bcs.u8(),
		certified_epoch: bcs.option(bcs.u32()),
		storage: Storage,
		deletable: bcs.bool()
	}
});
const BlobIdDerivation = new MoveStruct({
	name: `${$moduleName}::BlobIdDerivation`,
	fields: {
		encoding_type: bcs.u8(),
		size: bcs.u64(),
		root_hash: bcs.u256()
	}
});
/**
* Adds the metadata dynamic field to the Blob.
*
* Aborts if the metadata is already present.
*/
function addMetadata(options) {
	const packageAddress = options.package ?? "@local-pkg/walrus";
	const argumentsTypes = [null, null];
	const parameterNames = ["self", "metadata"];
	return (tx) => tx.moveCall({
		package: packageAddress,
		module: "blob",
		function: "add_metadata",
		arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
	});
}
/**
* Inserts a key-value pair into the metadata.
*
* If the key is already present, the value is updated. Creates new metadata on the
* Blob object if it does not exist already.
*/
function insertOrUpdateMetadataPair(options) {
	const packageAddress = options.package ?? "@local-pkg/walrus";
	const argumentsTypes = [
		null,
		"0x1::string::String",
		"0x1::string::String"
	];
	const parameterNames = [
		"self",
		"key",
		"value"
	];
	return (tx) => tx.moveCall({
		package: packageAddress,
		module: "blob",
		function: "insert_or_update_metadata_pair",
		arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
	});
}
/**
* Removes the metadata associated with the given key.
*
* Aborts if the metadata does not exist.
*/
function removeMetadataPair(options) {
	const packageAddress = options.package ?? "@local-pkg/walrus";
	const argumentsTypes = [null, "0x1::string::String"];
	const parameterNames = ["self", "key"];
	return (tx) => tx.moveCall({
		package: packageAddress,
		module: "blob",
		function: "remove_metadata_pair",
		arguments: normalizeMoveArguments(options.arguments, argumentsTypes, parameterNames)
	});
}

//#endregion
export { Blob, addMetadata, insertOrUpdateMetadataPair, removeMetadataPair };
