// @mysten/walrus/dist/contracts/walrus/bls_aggregate.mjs
import { MoveEnum, MoveStruct } from "../utils/index.mjs";
import { Element } from "./deps/sui/group_ops.mjs";
import { bcs } from "@mysten/sui/bcs";

//#region src/contracts/walrus/bls_aggregate.ts
/**************************************************************
* THIS FILE IS GENERATED AND SHOULD NOT BE MANUALLY MODIFIED *
**************************************************************/
const $moduleName = "@local-pkg/walrus::bls_aggregate";
const BlsCommitteeMember = new MoveStruct({
	name: `${$moduleName}::BlsCommitteeMember`,
	fields: {
		public_key: Element,
		weight: bcs.u16(),
		node_id: bcs.Address
	}
});
const BlsCommittee = new MoveStruct({
	name: `${$moduleName}::BlsCommittee`,
	fields: {
		members: bcs.vector(BlsCommitteeMember),
		n_shards: bcs.u16(),
		epoch: bcs.u32(),
		total_aggregated_key: Element
	}
});
/** The type of weight verification to perform. */
const RequiredWeight = new MoveEnum({
	name: `${$moduleName}::RequiredWeight`,
	fields: {
		Quorum: null,
		OneCorrectNode: null
	}
});

//#endregion
export { BlsCommittee };
