// @mysten/walrus/dist/contracts/walrus/event_blob.mjs
import { MoveStruct } from "../utils/index.mjs";
import { VecMap } from "./deps/sui/vec_map.mjs";
import { bcs } from "@mysten/sui/bcs";

//#region src/contracts/walrus/event_blob.ts
/**************************************************************
* THIS FILE IS GENERATED AND SHOULD NOT BE MANUALLY MODIFIED *
**************************************************************/
/** Module to certify event blobs. */
const $moduleName = "@local-pkg/walrus::event_blob";
const EventBlobAttestation = new MoveStruct({
	name: `${$moduleName}::EventBlobAttestation`,
	fields: {
		checkpoint_sequence_num: bcs.u64(),
		epoch: bcs.u32()
	}
});
const EventBlob = new MoveStruct({
	name: `${$moduleName}::EventBlob`,
	fields: {
		blob_id: bcs.u256(),
		ending_checkpoint_sequence_number: bcs.u64()
	}
});
const EventBlobCertificationState = new MoveStruct({
	name: `${$moduleName}::EventBlobCertificationState`,
	fields: {
		latest_certified_blob: bcs.option(EventBlob),
		aggregate_weight_per_blob: VecMap(EventBlob, bcs.u16())
	}
});

//#endregion
export { EventBlobAttestation, EventBlobCertificationState };
