// @mysten/walrus/dist/contracts/walrus/staking_pool.mjs
import { MoveEnum, MoveStruct } from "../utils/index.mjs";
import { StorageNodeInfo } from "./storage_node.mjs";
import { PendingValues } from "./pending_values.mjs";
import { Table } from "./deps/sui/table.mjs";
import { Balance } from "./deps/sui/balance.mjs";
import { Authorized } from "./auth.mjs";
import { Bag } from "./deps/sui/bag.mjs";
import { bcs } from "@mysten/sui/bcs";

//#region src/contracts/walrus/staking_pool.ts
/**************************************************************
* THIS FILE IS GENERATED AND SHOULD NOT BE MANUALLY MODIFIED *
**************************************************************/
/** Module: staking_pool */
const $moduleName = "@local-pkg/walrus::staking_pool";
const VotingParams = new MoveStruct({
	name: `${$moduleName}::VotingParams`,
	fields: {
		storage_price: bcs.u64(),
		write_price: bcs.u64(),
		node_capacity: bcs.u64()
	}
});
/** Represents the state of the staking pool. */
const PoolState = new MoveEnum({
	name: `${$moduleName}::PoolState`,
	fields: {
		Active: null,
		Withdrawing: bcs.u32(),
		Withdrawn: null
	}
});
const StakingPool = new MoveStruct({
	name: `${$moduleName}::StakingPool`,
	fields: {
		id: bcs.Address,
		state: PoolState,
		voting_params: VotingParams,
		node_info: StorageNodeInfo,
		activation_epoch: bcs.u32(),
		latest_epoch: bcs.u32(),
		wal_balance: bcs.u64(),
		num_shares: bcs.u64(),
		pending_shares_withdraw: PendingValues,
		pre_active_withdrawals: PendingValues,
		pending_commission_rate: PendingValues,
		commission_rate: bcs.u16(),
		exchange_rates: Table,
		pending_stake: PendingValues,
		rewards_pool: Balance,
		commission: Balance,
		commission_receiver: Authorized,
		governance_authorized: Authorized,
		extra_fields: Bag
	}
});

//#endregion
export { StakingPool };
