// @mysten/walrus/dist/contracts/walrus/staking_inner.mjs
import { MoveEnum, MoveStruct } from "../utils/index.mjs";
import { ObjectTable } from "./deps/sui/object_table.mjs";
import { ExtendedField } from "./extended_field.mjs";
import { Committee } from "./committee.mjs";
import { EpochParams } from "./epoch_parameters.mjs";
import { bcs } from "@mysten/sui/bcs";

//#region src/contracts/walrus/staking_inner.ts
/**************************************************************
* THIS FILE IS GENERATED AND SHOULD NOT BE MANUALLY MODIFIED *
**************************************************************/
const $moduleName = "@local-pkg/walrus::staking_inner";
/** The epoch state. */
const EpochState = new MoveEnum({
	name: `${$moduleName}::EpochState`,
	fields: {
		EpochChangeSync: bcs.u16(),
		EpochChangeDone: bcs.u64(),
		NextParamsSelected: bcs.u64()
	}
});
const StakingInnerV1 = new MoveStruct({
	name: `${$moduleName}::StakingInnerV1`,
	fields: {
		n_shards: bcs.u16(),
		epoch_duration: bcs.u64(),
		first_epoch_start: bcs.u64(),
		pools: ObjectTable,
		epoch: bcs.u32(),
		active_set: ExtendedField,
		next_committee: bcs.option(Committee),
		committee: Committee,
		previous_committee: Committee,
		next_epoch_params: bcs.option(EpochParams),
		epoch_state: EpochState,
		next_epoch_public_keys: ExtendedField
	}
});

//#endregion
export { StakingInnerV1 };
