// @mysten/walrus/dist/contracts/walrus/storage_node.mjs
import { MoveStruct } from "../utils/index.mjs";
import { ExtendedField } from "./extended_field.mjs";
import { Element } from "./deps/sui/group_ops.mjs";
import { EventBlobAttestation } from "./event_blob.mjs";
import { bcs } from "@mysten/sui/bcs";

//#region src/contracts/walrus/storage_node.ts
/**************************************************************
* THIS FILE IS GENERATED AND SHOULD NOT BE MANUALLY MODIFIED *
**************************************************************/
const $moduleName = "@local-pkg/walrus::storage_node";
const StorageNodeInfo = new MoveStruct({
	name: `${$moduleName}::StorageNodeInfo`,
	fields: {
		name: bcs.string(),
		node_id: bcs.Address,
		network_address: bcs.string(),
		public_key: Element,
		next_epoch_public_key: bcs.option(Element),
		network_public_key: bcs.vector(bcs.u8()),
		metadata: ExtendedField
	}
});
const StorageNodeCap = new MoveStruct({
	name: `${$moduleName}::StorageNodeCap`,
	fields: {
		id: bcs.Address,
		node_id: bcs.Address,
		last_epoch_sync_done: bcs.u32(),
		last_event_blob_attestation: bcs.option(EventBlobAttestation),
		deny_list_root: bcs.u256(),
		deny_list_sequence: bcs.u64(),
		deny_list_size: bcs.u64()
	}
});

//#endregion
export { StorageNodeInfo };
