// @mysten/walrus/dist/contracts/walrus/system_state_inner.mjs
import { MoveStruct } from "../utils/index.mjs";
import { ExtendedField } from "./extended_field.mjs";
import { EventBlobCertificationState } from "./event_blob.mjs";
import { BlsCommittee } from "./bls_aggregate.mjs";
import { FutureAccountingRingBuffer } from "./storage_accounting.mjs";
import { bcs } from "@mysten/sui/bcs";

//#region src/contracts/walrus/system_state_inner.ts
/**************************************************************
* THIS FILE IS GENERATED AND SHOULD NOT BE MANUALLY MODIFIED *
**************************************************************/
const $moduleName = "@local-pkg/walrus::system_state_inner";
const SystemStateInnerV1 = new MoveStruct({
	name: `${$moduleName}::SystemStateInnerV1`,
	fields: {
		committee: BlsCommittee,
		total_capacity_size: bcs.u64(),
		used_capacity_size: bcs.u64(),
		storage_price_per_unit_size: bcs.u64(),
		write_price_per_unit_size: bcs.u64(),
		future_accounting: FutureAccountingRingBuffer,
		event_blob_certification_state: EventBlobCertificationState,
		deny_list_sizes: ExtendedField
	}
});

//#endregion
export { SystemStateInnerV1 };
